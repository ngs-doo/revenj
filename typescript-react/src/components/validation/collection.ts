import { removeSpaces } from '../../util/StringUtils/StringUtils';
import { validatorCreatorFactory } from './validatorCreatorFactory';

export const maxLengthCreator = (
  maxLength: number,
  operator?: (value: number) => (input: string | any[] | number) => boolean) =>
    // Number should _not_ be here, but it happens in real life and will have to stay until we have
    // typesafe forms across the board
    validatorCreatorFactory<number, string | any[] | number, any, any>({
      getValidatorBaseValue: () => maxLength,
      getValidatorErrorMessage: (value) => `Maximum length of field is ${value}`,
      operator: operator || ((value) => (input) =>
        value == null || (typeof input === 'number' ? String(input).length <= value : input.length <= value)),
    });

export const maxDecimalLengthCreator = (
  maxLength: number,
  operator?: (value: number) => (input: string | number) => boolean) =>
    validatorCreatorFactory<number, string | number, any, any>({
      getValidatorBaseValue: () => maxLength,
      getValidatorErrorMessage: (value) => `Maximum number of decimals is: ${value}`,
      operator: operator || ((value) => {
        const regex = new RegExp(`^-?[0-9]*\.[0-9]{0,${value}}$`);
        return (input) => regex.test(String(input));
    })});

export const exactLengthCreator = (
  expectedLength: number,
  operator?: (value: number) => (input: string | any[]) => boolean) =>
    validatorCreatorFactory<number, string | any[], any, any>({
      getValidatorBaseValue: () => expectedLength,
      getValidatorErrorMessage: (value) => `Required length of field is exactly ${value}`,
      operator: operator || ((value) => (input) => input == null || input.length === value),
    });

export const minLengthCreator = (minLength: number) =>
  validatorCreatorFactory<number, number | string, any, any>({
    getValidatorBaseValue: () => minLength,
    getValidatorErrorMessage: (value) => `Minimum length of field is ${value}`,
    operator: (minLength) => (input) =>
      input == null || (String(input).length >= minLength),
  });

export const maxLengthCreatorWithIgnoreByRegex = (maxLength: number, regex: RegExp) => maxLengthCreator(
    maxLength,
    (
      (value) => (input) =>
      value == null ||
      (typeof input === 'number'
        ? String(input).replace(regex, '').length <= value
        : Array.isArray(input)
          ? input.length <= value
          : input.replace(regex, '').length <= value
      )
    ),
  );

export const minQueryLengthCreator = (minLength: number = 3) =>
  validatorCreatorFactory<number, string, any, any>({
    getValidatorBaseValue: () => minLength,
    getValidatorErrorMessage: (value) => `Minimum query length is ${value}`,
    operator: (minLength) => (input) =>
      input == null || (removeSpaces(input).length >= minLength),
  });

export const isNotInCreator = (values: any[]) =>
  validatorCreatorFactory<any[], any, any, any>({
    getValidatorBaseValue: () => values,
    getValidatorErrorMessage: (value) => value.length > 1
      ? `Value must not be equal to any of: ${value.join(', ')}`
      : `Value must not be equal to ${value}`,
    operator: (values) => (input) => values.indexOf(input) === -1,
  });

export const isInCreator = (values: any[]) =>
  validatorCreatorFactory<any[], any, any, any>({
    getValidatorBaseValue: () => values,
    getValidatorErrorMessage: (value) => `Value must not be equal to ${value}`,
    operator: (values) => (input) => values.indexOf(input) !== -1,
  });
