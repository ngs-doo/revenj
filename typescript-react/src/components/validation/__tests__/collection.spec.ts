import { exactLengthCreator, isInCreator, isNotInCreator, maxLengthCreator, maxDecimalLengthCreator, minLengthCreator } from '../collection';

describe('validation', () => {
  describe('maxLength', () => {
    it('should return undefined if input value is not longer than maxLength', () => {
      const maxLength20 = maxLengthCreator(20)();
      expect(maxLength20('under 20 chars', {})).toBeUndefined();
      expect(maxLength20('exactly twenty chars', {})).toBeUndefined();
    });

    it('should return an error message if input value is longer than maxLength', () => {
      const maxLength20 = maxLengthCreator(20)();
      expect(maxLength20('slightly over twenty characters', {})).toBeDefined();
      expect(maxLength20('slightly over twenty characters', {})).toMatchSnapshot();
    });

    it('should use the digit count as length for numbers (regression test for IN-8549)', () => {
      const maxLength3 = maxLengthCreator(3)();
      expect(maxLength3(1, {})).not.toBeDefined();
      expect(maxLength3(1234, {})).toBeDefined();
      expect(maxLength3(1234, {})).toMatchSnapshot();
    });
  });

  describe('maxDecimalLength', () => {
    it('should return undefined if input value is not longer than maxLength', () => {
      const maxDecimalLength10 = maxDecimalLengthCreator(10)();
      expect(maxDecimalLength10('12.01234567', {})).toBeUndefined();
      expect(maxDecimalLength10('1.0123456789', {})).toBeUndefined();
      expect(maxDecimalLength10(1.0123456789, {})).toBeUndefined();
      expect(maxDecimalLength10('100', {})).toBeUndefined();
    });

    it('should return an error message if string input value is longer than maxDecimalLength', () => {
      const maxDecimalLength10 = maxDecimalLengthCreator(10)();
      expect(maxDecimalLength10('1.012345678901', {})).toBeDefined();
      expect(maxDecimalLength10('1.012345678901', {})).toMatchSnapshot();
    });

    it('should return an error message if numeric input value is longer than maxDecimalLength', () => {
      const maxDecimalLength10 = maxDecimalLengthCreator(10)();
      expect(maxDecimalLength10(1.012345678901, {})).toBeDefined();
      expect(maxDecimalLength10(1.012345678901, {})).toMatchSnapshot();
    });

    it('should not allow decimal numbers if maxDecimalLength is 0', () => {
      const maxDecimalLength0 = maxDecimalLengthCreator(0)();
      expect(maxDecimalLength0(1, {})).toBeUndefined();
      expect(maxDecimalLength0('1.0', {})).toBeDefined();
      expect(maxDecimalLength0('1.0', {})).toMatchSnapshot();
    });
  });

  describe('exactLength', () => {
    it('should return undefined if the value is null, empty string, or has the exact length', () => {
      const exactLength5 = exactLengthCreator(5)();
      expect(exactLength5(null as any, {})).toBeUndefined();
      expect(exactLength5(undefined as any, {})).toBeUndefined();
      expect(exactLength5('', {})).toBeUndefined();
      expect(exactLength5('12345', {})).toBeUndefined();
      expect(exactLength5([1, 2, 3, 4, 5], {})).toBeUndefined();
    });

    it('should return an error message if the input has a different length', () => {
      const exactLength5 = exactLengthCreator(5)();
      expect(exactLength5('2', {})).toBeDefined();
      expect(exactLength5('2', {})).toMatchSnapshot();
      expect(exactLength5('333333333', {})).toBeDefined();
      expect(exactLength5('333333333', {})).toMatchSnapshot();
      expect(exactLength5([], {})).toBeDefined();
      expect(exactLength5([], {})).toMatchSnapshot();
      expect(exactLength5([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], {})).toBeDefined();
      expect(exactLength5([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], {})).toMatchSnapshot();
    });
  });

  describe('minLength', () => {
    it('should return undefined if input value is longer than minLength', () => {
      const minLength10 = minLengthCreator(10)();
      expect(minLength10('more than 10 chars', {})).toBeUndefined();
      expect(minLength10('five chars', {})).toBeUndefined();
    });

    it('should return an error message if input value is shorter than minLength', () => {
      const minLength7 = minLengthCreator(7)();
      expect(minLength7('4-char', {})).toBeDefined();
      expect(minLength7('4-char', {})).toMatchSnapshot();
    });

    it('should use the digit count as length for numbers (regression test for IN-8549)', () => {
      const minLength3 = minLengthCreator(3)();
      expect(minLength3(134543, {})).not.toBeDefined();
      expect(minLength3(1, {})).toBeDefined();
      expect(minLength3(12, {})).toMatchSnapshot();
    });
  });

  describe('isIn', () => {
    it('should return undefined if input value is found in the set of values', () => {
      const isIn = isInCreator([3, 5, 6])();
      expect(isIn(3, {})).toBeUndefined();
      expect(isIn(6, {})).toBeUndefined();
    });

    it('should return an error message if input value is not found in the set of values', () => {
      const isIn = isInCreator([3, 5, 6])();
      expect(isIn(1, {})).toBeDefined();
      expect(isIn(9, {})).toBeDefined();
    });
  });

  describe('isNotIn', () => {
    it('should return undefined if input value is not found in the set of values', () => {
      const isIn = isNotInCreator([3, 5, 6])();
      expect(isIn(1, {})).toBeUndefined();
      expect(isIn(9, {})).toBeUndefined();
    });

    it('should return an error message if input value is found in the set of values', () => {
      const isIn = isNotInCreator([3, 5, 6])();
      expect(isIn(3, {})).toBeDefined();
      expect(isIn(6, {})).toBeDefined();
    });
  });
});
