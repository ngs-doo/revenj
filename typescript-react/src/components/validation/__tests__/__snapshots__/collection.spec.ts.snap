// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation exactLength should return an error message if the input has a different length 1`] = `"Required length of field is exactly 5"`;

exports[`validation exactLength should return an error message if the input has a different length 2`] = `"Required length of field is exactly 5"`;

exports[`validation exactLength should return an error message if the input has a different length 3`] = `"Required length of field is exactly 5"`;

exports[`validation exactLength should return an error message if the input has a different length 4`] = `"Required length of field is exactly 5"`;

exports[`validation maxDecimalLength should not allow decimal numbers if maxDecimalLength is 0 1`] = `"Maximum number of decimals is: 0"`;

exports[`validation maxDecimalLength should return an error message if numeric input value is longer than maxDecimalLength 1`] = `"Maximum number of decimals is: 10"`;

exports[`validation maxDecimalLength should return an error message if string input value is longer than maxDecimalLength 1`] = `"Maximum number of decimals is: 10"`;

exports[`validation maxLength should return an error message if input value is longer than maxLength 1`] = `"Maximum length of field is 20"`;

exports[`validation maxLength should use the digit count as length for numbers (regression test for IN-8549) 1`] = `"Maximum length of field is 3"`;

exports[`validation minLength should return an error message if input value is shorter than minLength 1`] = `"Minimum length of field is 7"`;

exports[`validation minLength should use the digit count as length for numbers (regression test for IN-8549) 1`] = `"Minimum length of field is 3"`;
